### PR TITLE
Add coding_guideline.rst in odemis developer manual

### PR DESCRIPTION
**Pull request by K4rishma**
Coding guidelines document is created from meeting notes of software knowledge transfer and comments received in [meeting notes ](https://docs.google.com/document/d/1y5fdmL5cgSKknFq4HA27ZUgzgrF1dujEwLiXEceqBOE/edit)and [this doc](https://docs.google.com/document/d/1sH_p89dpFfDtLFJj_s-CGMIoFYqg3eSfIj81SD8sn4Q/edit).

For now I have added the coding guidelines chapter at the top of the manual. Please let me know if that is the correct order.

Kindly review this document. If any typos/grammar/ are visible, please add those too in the review. 

Output :
- compiled using 
- `make latexpdf` `make html`

### CODE REVIEW

File: doc/develop/coding-guidelines.rst [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+
+ self.pixelSize = model.VigilantAttribute(pxs, unit="m", readonly=True)
+ self.horizontalFoV = model.FloatContinuous(hfv, range=[10e-9, 10e-3],
+                                               unit="m")
+ self.magnification = model.VigilantAttribute(self._hfw_nomag / hfv,
+                                                 unit="", readonly=True)
+
+* Naming of test files follow the below ways to provide consistency for files in test folders (no _init_.py file), which are included in each sub-directories of src folder.
+
+ * A unittest convention is used to name the test class for a module as TestModuleName(unittest.Testcase).
```
Comment (nandishjpatel) [inline doc/develop/coding-guidelines.rst:41]: @tepals, @pieleric should we also have a unit testcase naming convention? [example UnitOfWork_StateUnderTest_ExpectedBehavior](https://osherove.com/blog/2005/4/3/naming-standards-for-unit-tests.html)

File: doc/develop/coding-guidelines.rst [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+
+The usage of public attributes is useful when their access is required by a third
+party or another class. In this scenario, public attributes can be used.
+
+For e.g. 1.2 the implementation of public and private attributes
+
+.. code-block:: python
+
+    class Pokemon:
+        """
```
Comment (nandishjpatel) [inline doc/develop/coding-guidelines.rst:72]:
```suggestion
        """Links special powers to the given pokemon type."""
```

For single-lined description, should `"""` be on the same line?

File: doc/develop/coding-guidelines.rst [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+
+For e.g. 2.3 the test class should be defined as
+
+.. code-block:: python
+
+    class TestOrsayStatic(unittest.TestCase):
+        """
+        Tests which don't need an Orsay component ready
+        """
+
```
Comment (K4rishma) [inline doc/develop/coding-guidelines.rst:140]: @nandishjpatel  in regards to having quotes in the same line, I saw some code in the repository like this. Note sure what is the preferred method.

File: doc/develop/coding-guidelines.rst [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+Within each of the three sub categories follow the alphabetical order. “From” and “import” statements are mixed.
+
+When importing many functions or variables from a single module, the length of the import line may not fit within the maximum line length. In that case, use parentheses to place the import over multiple lines.
+
+For e.g. 4.1 standard and internal imports
+
+.. code-block:: python
+
+    from collections.abc import Iterable, Mapping
```
Comment (sneepstefan) [inline doc/develop/coding-guidelines.rst:201]: Shouldn't this line come after import cairo?

File: doc/develop/coding-guidelines.rst (hunk 1 above)
Comment (pieleric) [reply doc/develop/coding-guidelines.rst:41]: Why not, it sounds a pretty good idea... but for now we have no convention, so it'll not be very much respected for all the current test cases.

File: doc/develop/coding-guidelines.rst [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+    class LakeshoreBaseTest:
+        """Base class for testing different models of LakeShore temperature controller."""
+
+        @classmethod
+        def tearDownClass(cls):
+            cls.dev.terminate()  # free up socket.
+
+In class, mark the attributes as private with a first letter as _. They are
+accessible normally from other classes, but by convention, as developers we know that they are
+private to the classes and should not be externally modified.
```
Comment (pieleric) [inline doc/develop/coding-guidelines.rst:60]: I'd avoid "as developers", and actually it should not even be read. So I'd rephrase like:
"Although technically Python still allows to access them from other classes, by convention, such attributes should not be read or modified externally."

File: doc/develop/coding-guidelines.rst [hunk 6]
```diff
@@ -0,0 +1,213 @@
...
+If there are public attributes, the attributes might have to be changed in several
+files and folders.
```
Comment (pieleric) [inline doc/develop/coding-guidelines.rst:62]: I'd reverse it, with something like "Private attributes can be changed or removed later, without having to look at the code outside of the class they belong to."

File: doc/develop/coding-guidelines.rst [hunk 7]
```diff
@@ -0,0 +1,213 @@
...
+    def get_equality(arg1, arg2):
+        """Summary line.
+
+        Extended description of function.
+
+        :param arg1: (int) Description of arg1.
+        :param arg2: (str) Description of arg2.
+        :raise: ValueError if arg1 is equal to arg2
+        :return: (bool) Description of return value
+
```
Comment (pieleric) [inline doc/develop/coding-guidelines.rst:102]: There is no need for a empty line here.

File: doc/develop/coding-guidelines.rst [hunk 8]
```diff
@@ -0,0 +1,213 @@
...
+Docstrings
+==================
+
+Use three double quotes for the docstring. The docstring consists of function definition, description of input and return arguments. We will follow the reStructuredText style to write the docstring.
+For e.g. 2.1 when type is declared in the docstring
+
+.. code-block:: python
+
+    def get_equality(arg1, arg2):
+        """Summary line.
```
Comment (pieleric) [inline doc/develop/coding-guidelines.rst:94]: Let's put the summary line as a separate line from the `"""`.

File: doc/develop/coding-guidelines.rst (hunk 2 above)
Comment (pieleric) [reply doc/develop/coding-guidelines.rst:72]: I don't have strong feelings about it. We could say that both options are allowed. If we want to reduce alternatives, I'd suggest to go simple and say we should always have separate lines (this way, if extending the docstring after, there are less changes).

File: doc/develop/coding-guidelines.rst [hunk 9]
```diff
@@ -0,0 +1,213 @@
...
+            with self.assertRaises(HwError):
+                orsay.OrsayComponent(name="Orsay", role="orsay", host="", children="")
+
+
+
+General Guidelines
+
+* For modules, add a license docstring (i.e. within three double quotes) at the top.
+
+  * Exception: It is not necessary for '__init__.py' to carry a license.
```
Comment (pieleric) [inline doc/develop/coding-guidelines.rst:168]: "It is not necessary for **empty**  `'__init__.py'` to carry a license."

File: doc/develop/coding-guidelines.rst [hunk 10]
```diff
@@ -0,0 +1,213 @@
...
+
+* Add the description of the script as a comment below the license.
+
+* Add a docstring to define the class. See example 1.1.
+
+Code Cleaning up
+==================
+
+* If you clean the code for the current feature, make a separate commit for cleaning. The first commit  is recommended for clean-up in the working branch.
```
Comment (pieleric) [inline doc/develop/coding-guidelines.rst:177]: "The first commit  is recommended for clean-up in the working branch." -> "It is recommended to place this commit as the first commit of a pull-request."

File: doc/develop/coding-guidelines.rst [hunk 11]
```diff
@@ -0,0 +1,213 @@
...
+
+Use the following order while importing python modules
+
+* Standard library: See: https://docs.python.org/3/library/
+
+* External: (aka 3rd party) library: e.g. numpy, pandas etc.
+
+* Internal: For e.g. odemis packages.
+
+Within each of the three sub categories follow the alphabetical order. “From” and “import” statements are mixed.
```
Comment (pieleric) [inline doc/develop/coding-guidelines.rst:193]: Drop the upper case from "From". I'd suggest even to mark these words as fixed-width, by placing them around back-ticks.

File: doc/develop/coding-guidelines.rst (hunk 4 above)
Comment (pieleric) [reply doc/develop/coding-guidelines.rst:201]: No, because it's standard lib. However, there should be an empty line between that line and cairo.

File: doc/develop/coding-guidelines.rst [hunk 12]
```diff
@@ -0,0 +1,213 @@
...
+
+When importing many functions or variables from a single module, the length of the import line may not fit within the maximum line length. In that case, use parentheses to place the import over multiple lines.
+
+For e.g. 4.1 standard and internal imports
+
+.. code-block:: python
+
+    from collections.abc import Iterable, Mapping
+    import cairo
+    import logging
```
Comment (pieleric) [inline doc/develop/coding-guidelines.rst:203]: `import logging` is from the standard lib, so that should go to the first group.

File: doc/develop/index.rst [hunk 13]
```diff
@@ -15,6 +15,7 @@ This document covers the main aspects of development with Odemis.
 .. toctree::
     :maxdepth: 2
 
+    coding-guidelines
     sys-arch
     extending
```
Comment (pieleric) [inline doc/develop/index.rst:20]: I don't think it's nice for someone who want to starts with learning about Odemis to be bothered by every details of coding style. Let's place it after "extending".

File: doc/develop/coding-guidelines.rst [hunk 14]
```diff
@@ -0,0 +1,213 @@
...
+* Function name should be in snake_case in which the multiple words in small capitalization are joined with underscore where the first word is verb. For e.g. get_image()
+
+* Import modules should not be renamed. Use the original import name. For e.g. import numpy (and not import numpy as np)
+
+ * Exception: import matplotlib.pyplot as plt.
+
+* VAs in odemis should be in mixedCase which is different from PEP 8.
+
+
+Exception : If dependent on outer dependencies which follow different
```
Comment (pieleric) [inline doc/develop/coding-guidelines.rst:23]: No need for a space between exception and `:`.

File: doc/develop/coding-guidelines.rst [hunk 15]
```diff
@@ -0,0 +1,213 @@
...
+developer for odemis.
+
+Naming Conventions
+==================
+
+* Class names should be in CamelCase in which multiple words are connected by capitalizing the first letter of the word. For e.g. MyClass().
+
+* Function name should be in snake_case in which the multiple words in small capitalization are joined with underscore where the first word is verb. For e.g. get_image()
+
+* Import modules should not be renamed. Use the original import name. For e.g. import numpy (and not import numpy as np)
```
Comment (pieleric) [inline doc/develop/coding-guidelines.rst:16]: Add backticks around `import numpy` and also around `import numpy as np`.

File: doc/develop/coding-guidelines.rst [hunk 16]
```diff
@@ -0,0 +1,213 @@
...
+
+        Extended description of function.
+
+        :param arg1: (int) Description of arg1.
+        :param arg2: (str) Description of arg2.
+        :raise: ValueError if arg1 is equal to arg2
+        :return: (bool) Description of return value
+
+        """
+
```
Comment (pieleric) [inline doc/develop/coding-guidelines.rst:104]: Do we want an empty line after the docstring? I would tend to say "no". So let's drop that line.

**Review by pieleric — CHANGES_REQUESTED** [review]

**Review by pieleric — CHANGES_REQUESTED** [review]

File: doc/develop/coding-guidelines.rst [hunk 17]
```diff
@@ -0,0 +1,212 @@
...
+    import logging
+
+    import numpy
+
+    from odemis.gui import FG_COLOUR_DIS
+    from odemis.gui.comp.buttons import ImageToggleButton
+    from odemis.model import (MD_AT_AR, MD_AT_CL, MD_AT_EK, MD_AT_EM, MD_AT_FLUO,
+                              MD_AT_SLIT, MD_AT_SPECTRUM, MD_AT_TEMPORAL,
+                              MD_AT_TEMPSPECTRUM, MD_AT_FIB)
+    import odemis
```
[the remark this answers is not in the document]
Comment (tepals) [reply doc/develop/coding-guidelines.rst:210]: I thought we decided we would group them, but that's indeed not what's written on line 191

[the remark this answers is not in the document]
Comment (pieleric) [reply doc/develop/coding-guidelines.rst]: I'm also fine with it. Apparently that's the default in PyCharm, so that simplifies quite a few settings. Is "import" after "from" or the opposite?

[the remark this answers is not in the document]
Comment (tepals) [reply doc/develop/coding-guidelines.rst]: "import" is before "from", so:
```python
import odemis
from odemis import model
```

[the remark this answers is not in the document]
Comment (K4rishma) [reply doc/develop/coding-guidelines.rst]: Are the current changes in the import according to the agreement?

File: doc/develop/coding-guidelines.rst [hunk 18]
```diff
@@ -0,0 +1,212 @@
...
+
+.. code-block:: python
+
+      import cairo
+      from collections.abc import Iterable, Mapping
+      import logging
+
+      import numpy
+
+      import odemis
```
Comment (tepals) [inline doc/develop/coding-guidelines.rst:205]: this one should go below `from odemis.model import (.....................`

**Review by BasHolweg — APPROVED** [review]

File: doc/develop/coding-guidelines.rst [hunk 19]
```diff
@@ -0,0 +1,212 @@
...
+
+Within each of the three sub categories follow the alphabetical order. ``from`` and ``import`` statements are mixed.
+
+When importing many functions or variables from a single module, the length of the import line may not fit within the maximum line length. In that case, use parentheses to place the import over multiple lines.
+
+For e.g. 4.1 standard and internal imports
+
+.. code-block:: python
+
+      import cairo
```
Comment (pieleric) [inline doc/develop/coding-guidelines.rst:199]: Cairo is an external library => 2nd block

File: doc/develop/coding-guidelines.rst [hunk 20]
```diff
@@ -0,0 +1,212 @@
...
+      from collections.abc import Iterable, Mapping
+      import logging
```
Comment (pieleric) [inline doc/develop/coding-guidelines.rst:201]: From should be after import => swap lines.

File: doc/develop/coding-guidelines.rst (hunk 18 above)
Comment (pieleric) [reply doc/develop/coding-guidelines.rst:205]: No. All the from's should be after all the imports. So `import odemis.util.units as units` should come here.

File: doc/develop/coding-guidelines.rst [hunk 21]
```diff
@@ -0,0 +1,212 @@
...
+
+Use the following order while importing python modules
+
+* Standard library: See: https://docs.python.org/3/library/
+
+* External: (aka 3rd party) library: e.g. numpy, pandas etc.
+
+* Internal: For e.g. odemis packages.
+
+Within each of the three sub categories follow the alphabetical order. ``from`` and ``import`` statements are mixed.
```
Comment (pieleric) [inline doc/develop/coding-guidelines.rst:191]: "In each block, all the ``import`` statements are first, then come all the ``from`` statements."

**Review by pieleric — CHANGES_REQUESTED** [review]

**Comment by K4rishma** [comment]
@pieleric @tepals requestes changes added

File: doc/develop/coding-guidelines.rst [hunk 22]
```diff
@@ -0,0 +1,214 @@
...
+      import cairo
+      import numpy
+
+      import odemis
+      import odemis.util.units as units
+      from odemis.gui import FG_COLOUR_DIS
+      from odemis.gui.comp.buttons import ImageToggleButton
+      from odemis.model import (MD_AT_AR, MD_AT_CL, MD_AT_EK, MD_AT_EM, MD_AT_FLUO,
+                              MD_AT_SLIT, MD_AT_SPECTRUM, MD_AT_TEMPORAL,
+                              MD_AT_TEMPSPECTRUM, MD_AT_FIB)
```
Comment (tepals) [inline doc/develop/coding-guidelines.rst:212]: It would be nice (and I think it's somewhere in PEP8) to align the first words on each row:

```suggestion
      from odemis.model import (MD_AT_AR, MD_AT_CL, MD_AT_EK, MD_AT_EM, MD_AT_FLUO,
                                MD_AT_SLIT, MD_AT_SPECTRUM, MD_AT_TEMPORAL,
                                MD_AT_TEMPSPECTRUM, MD_AT_FIB)
```

**Review by tepals — COMMENTED** [review]
Looks good! One small comment left

**Review by FCanberk — APPROVED** [review]

**Review by tepals — APPROVED** [review]
